File: phone-exchange-2-python/REST/pe2_response.py

from asyncio.windows_events import NULL
from dataclasses import dataclass
from typing import Optional, List, Type

@dataclass
class Error:
    Desc: Optional[str] = None
    TypeCode: Optional[str] = None
    DescCode: Optional[str] = None
    Type: Optional[str] = None

    def __str__(self) -> str:
        return f"Error: Desc={self.Desc}, TypeCode={self.TypeCode}, DescCode={self.DescCode}, Type={self.Type}"

@dataclass
class PortedInfo:
    OriginalName: Optional[str] = None
    OriginalLineType: Optional[str] = None
    PortedDate: Optional[str] = None
    LATA: Optional[str] = None

    def __str__(self) -> str:
        return (f"PortedInfo: OriginalName={self.OriginalName}, OriginalLineType={self.OriginalLineType}, "
                f"PortedDate={self.PortedDate}, LATA={self.LATA}")

@dataclass
class ExchangeInfo:
    PhoneNumber: Optional[str] = None
    Name: Optional[str] = None
    City: Optional[str] = None
    State: Optional[str] = None
    Country: Optional[str] = None
    LineType: Optional[str] = None
    TimeZone: Optional[str] = None
    Latitude: Optional[str] = None
    Longitude: Optional[str] = None
    SMSAddress: Optional[str] = None
    MMSAddress: Optional[str] = None
    PortedInfo: Optional['PortedInfo'] = None
    NoteCodes: Optional[str] = None
    NoteDescriptions: Optional[str] = None

    def __post_init__(self):
        if self.NoteCodes is None:
            self.NoteCodes = ""
        if self.NoteDescriptions is None:
            self.NoteDescriptions = ""

    def __str__(self) -> str:
        return (f"ExchangeInfo: PhoneNumber={self.PhoneNumber}, Name={self.Name}, City={self.City}, "
                f"State={self.State}, Country={self.Country}, LineType={self.LineType}, "
                f"TimeZone={self.TimeZone}, Latitude={self.Latitude}, Longitude={self.Longitude}, "
                f"SMSAddress={self.SMSAddress}, MMSAddress={self.MMSAddress}, "
                f"PortedInfo={self.PortedInfo}, NoteCodes=[{self.NoteCodes}], "
                f"NoteDescriptions={self.NoteDescriptions}")

@dataclass
class InternationalExchangeInfo:
    PhoneNumberIn: Optional[str] = None
    CountryCode: Optional[str] = None
    FormatNational: Optional[str] = None
    Extension: Optional[str] = None
    Locality: Optional[str] = None
    LocalityMatchLevel: Optional[str] = None
    TimeZone: Optional[str] = None
    Latitude: Optional[str] = None
    Longitude: Optional[str] = None
    Country: Optional[str] = None
    CountryISO2: Optional[str] = None
    CountryISO3: Optional[str] = None
    FormatInternational: Optional[str] = None
    FormatE164: Optional[str] = None
    Carrier: Optional[str] = None
    LineType: Optional[str] = None
    SMSAddress: Optional[str] = None
    MMSAddress: Optional[str] = None
    IsValid: bool = False
    IsValidForRegion: bool = False
    NoteCodes: Optional[str] = None
    NoteDescriptions: Optional[str] = None

    def __post_init__(self):
        if self.NoteCodes is None:
            self.NoteCodes = ""
        if self.NoteDescriptions is None:
            self.NoteDescriptions = ""

    def __str__(self) -> str:
        return (f"InternationalExchangeInfo: PhoneNumberIn={self.PhoneNumberIn}, CountryCode={self.CountryCode}, "
                f"FormatNational={self.FormatNational}, Extension={self.Extension}, Locality={self.Locality}, "
                f"LocalityMatchLevel={self.LocalityMatchLevel}, TimeZone={self.TimeZone}, "
                f"Latitude={self.Latitude}, Longitude={self.Longitude}, Country={self.Country}, "
                f"CountryISO2={self.CountryISO2}, CountryISO3={self.CountryISO3}, "
                f"FormatInternational={self.FormatInternational}, FormatE164={self.FormatE164}, "
                f"Carrier={self.Carrier}, LineType={self.LineType}, SMSAddress={self.SMSAddress}, "
                f"MMSAddress={self.MMSAddress}, IsValid={self.IsValid}, IsValidForRegion={self.IsValidForRegion}, "
                f"NoteCodes={self.NoteCodes}, NoteDescriptions={self.NoteDescriptions}")

@dataclass
class PE2Response:
    ExchangeInfoResults: Optional[List['ExchangeInfo']] = None
    InternationalExchangeInfo: Optional['InternationalExchangeInfo'] = None
    Error: Optional['Error'] = None

    def __post_init__(self):
        if self.ExchangeInfoResults is None:
            self.ExchangeInfoResults = []

    def __str__(self) -> str:
        exchange_info_str = '\n'.join(str(r) for r in self.ExchangeInfoResults) if self.ExchangeInfoResults else 'None'
        error_str = str(self.Error) if self.Error else 'None'
        return (f"PE2Response: ExchangeInfoResults=[\n{exchange_info_str}\n], "
                f"InternationalExchangeInfo=\n{self.InternationalExchangeInfo}\n, Error={error_str}")

File: phone-exchange-2-nodejs/REST/pe2_response.js
export class ExchangeInfo {
    constructor(data = {}) {
        this.PhoneNumber = data.PhoneNumber;
        this.Name = data.Name;
        this.City = data.City;
        this.State = data.State;
        this.Country = data.Country;
        this.LineType = data.LineType;
        this.TimeZone = data.TimeZone;
        this.Latitude = data.Latitude;
        this.Longitude = data.Longitude;
        this.SMSAddress = data.SMSAddress;
        this.MMSAddress = data.MMSAddress;
        this.PortedInfo = data.PortedInfo;
        this.NoteCodes = data.NoteCodes;
        this.NoteDescriptions = data.NoteDescriptions;
    }

    toString() {
        return `ExchangeInfo: PhoneNumber = ${this.PhoneNumber}, Name = ${this.Name}, City = ${this.City}, State = ${this.State}, Country = ${this.Country}, LineType = ${this.LineType}, TimeZone = ${this.TimeZone}, Latitude = ${this.Latitude}, Longitude = ${this.Longitude}, SMSAddress = ${this.SMSAddress}, MMSAddress = ${this.MMSAddress}, PortedInfo = ${this.PortedInfo}, NoteCodes = ${this.NoteCodes}, NoteDescriptions = ${this.NoteDescriptions}`;
    }
}

/**
 * Ported information for a phone number.
 */
export class PortedInfo {
    constructor(data = {}) {
        this.OriginalName = data.OriginalName;
        this.OriginalLineType = data.OriginalLineType;
        this.PortedDate = data.PortedDate;
        this.LATA = data.LATA;
    }

    toString() {
        return `PortedInfo: OriginalName = ${this.OriginalName}, OriginalLineType = ${this.OriginalLineType}, PortedDate = ${this.PortedDate}, LATA = ${this.LATA}`;
    }
}

/**
 * International phone exchange information for a validated phone number.
 */
export class InternationalExchangeInfo {
    constructor(data = {}) {
        this.PhoneNumberIn = data.NumberIn;
        this.CountryCode = data.CountryCode;
        this.FormatNational = data.FormatNational;
        this.Extension = data.Extension;
        this.Locality = data.Locality;
        this.LocalityMatchLevel = data.LocalityMatchLevel;
        this.TimeZone = data.TimeZone;
        this.Latitude = data.Latitude;
        this.Longitude = data.Longitude;
        this.Country = data.Country;
        this.CountryISO2 = data.CountryISO2;
        this.CountryISO3 = data.CountryISO3;
        this.FormatInternational = data.FormatInternational;
        this.FormatE164 = data.FormatE164;
        this.Carrier = data.Carrier;
        this.LineType = data.LineType;
        this.SMSAddress = data.SMSAddress;
        this.MMSAddress = data.MMSAddress;
        this.IsValid = data.IsValid !== undefined ? data.IsValid : null;
        this.IsValidForRegion = data.IsValidForRegion !== undefined ? data.IsValidForRegion : null;
        this.NoteCodes = data.NoteCodes;
        this.NoteDescriptions = data.NoteDescriptions;
    }

    toString() {
        return `InternationalExchangeInfo: PhoneNumberIn = ${this.PhoneNumberIn}, CountryCode = ${this.CountryCode}, FormatNational = ${this.FormatNational}, Extension = ${this.Extension}, Locality = ${this.Locality}, LocalityMatchLevel = ${this.LocalityMatchLevel}, TimeZone = ${this.TimeZone}, Latitude = ${this.Latitude}, Longitude = ${this.Longitude}, Country = ${this.Country}, CountryISO2 = ${this.CountryISO2}, CountryISO3 = ${this.CountryISO3}, FormatInternational = ${this.FormatInternational}, FormatE164 = ${this.FormatE164}, Carrier = ${this.Carrier}, LineType = ${this.LineType}, SMSAddress = ${this.SMSAddress}, MMSAddress = ${this.MMSAddress}, IsValid = ${this.IsValid}, IsValidForRegion = ${this.IsValidForRegion}, NoteCodes = ${this.NoteCodes}, NoteDescriptions = ${this.NoteDescriptions}`;
    }
}

/**
 * Error object for PE2 API responses.
 */
export class Error {
    constructor(data = {}) {
        this.Desc = data.Desc;
        this.TypeCode = data.TypeCode;
        this.DescCode = data.DescCode;
        this.Type = data.Type;
    }

    toString() {
        return `Error: Desc = ${this.Desc}, TypeCode = ${this.TypeCode}, DescCode = ${this.DescCode}, Type= ${this.Type}}`;
    }
}

/**
 * Response from PE2 GetExchangeInfo and GetInternationalExchangeInfo APIs, containing phone exchange information.
 */
export class PE2Response {
    constructor(data = {}) {
        this.ExchangeInfo = Array.isArray(data.ExchangeInfoResults)
            ? data.ExchangeInfoResults.map(info => new ExchangeInfo(info))
            : [];
        this.ExchangeInfoResults = (data.ExchangeInfoResults || []).map(info => new ExchangeInfo(info));
        this.InternationalExchangeInfo = data.InternationalExchangeInfo ? new InternationalExchangeInfo(data.InternationalExchangeInfo) : null;
        this.Error = data.Error ? new Error(data.Error) : null;
    }

    toString() {
        const exchangeInfoString = this.ExchangeInfoResults.length
            ? this.ExchangeInfoResults.map(info => info.toString()).join('; ')
            : 'null';
        const internationalInfoString = this.InternationalExchangeInfo.length
            ? this.InternationalExchangeInfo.map(info => info.toString()).join('; ')
            : 'null';
        return `PE2Response: ExchangeInfoResults = [${exchangeInfoString}], InternationalExchangeInfo = [${internationalInfoString}], Error = ${this.Error ? this.Error.toString() : 'null'}`;
    }
}

export default PE2Response;